csrc/optimization/consecutive_cast_pass.cpp

```diff
@@ -0,0 +1,151 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <ir/utils.h>
+#include <optimization/consecutive_cast_pass.h>
+
+namespace nvfuser::optimization {
+
+namespace {
+
+bool isSameDtypeCategory(const DataType& input_t, const DataType& output_t) {
+  if ((isIntegralType(input_t) && isIntegralType(output_t)) ||
+      (isFloatingPointType(input_t) && isFloatingPointType(output_t)) ||
+      (isComplexType(input_t) && isComplexType(output_t))) {
+    return true;
+  }
+  return false;
+}
+
+// note: returns
+//  - -1 : v0 contains strictly more information than v1;
+//  - 0  : a complex case, where each v0 and v1 isn't a super set of the other;
+//  - 1  : v0 and v1 has the same dtype;
+//  - 2  : v0 contains strictly less information than v1;
+int checkInformationLoss(Val* v0, Val* v1) {
```

Comment:
I think the code below would be much easier to follow with separate boolean functions for each case.